examples/job/adsorption-study.py

```diff
@@ -0,0 +1,268 @@
+        
+    def __repr__(self):
+        return(str(self.name))
+
+# Here, we've stored all of our jobs in a specific set. We'll run a Mongo Query to find jobs
+# that are inside of this set.
+job_set_id = "WLBD7M9dZBaNH2mMX"
+jobs = [JobData(job["_id"], job_endpoint, material_endpoint) for job in job_endpoint.list({"inSet":{"$elemMatch":{"_id":job_set_id}}})]
+
+
+# # 
+
+# # Download a Slab from the Platform
+# 
+# Now that we have our slab jobs, and a nice object to hold them, let's get them into ASE.
+
+# ## Get an ASE oject
+
+# In[]:
+
+
+for job in jobs:
+    job.filename = f"{'_'.join(map(str, job.miller_index))}_term{job.termination}.vasp"
+    download_contcar(job.jobId, job.job_endpoint, job.filename)
+    job.structure = ase.io.read(job.filename)
+
+
+# ## Generate the ASE Object
+
+# In[]:
+
+
+# Ensure we've got at least a 5x5 angstrom surface, to help combat lateral interactions in the adsorbates
+for job in jobs:
+    a, b, c = job.structure.cell.lengths()
+    while a < 5 or b < 5:
+        if a < 5:
+            job.structure *= [2,1,1]
+
+        if b < 5:
+            job.structure *= [1,2,1]
+        
+        a, b, c = job.structure.cell.lengths()
+
+
+# ## Find the Surface Sites
+
+# In[]:
+
+
+import ase.neighborlist
+
+def get_surface_indices(ase_structure, cns):
+    surface_indices = []
+    a, b, c = ase_structure.cell.lengths()
+    for index, (atom, cn) in enumerate(zip(ase_structure, cns)):
+        # Slabs are centered, so we can easily filter out one half of the atoms
+        if cn < 12 and atom.position[2] > c/2:
+            surface_indices.append(index)
+    return surface_indices
+
+for job in jobs:
+    src, dest = ase.neighborlist.neighbor_list("ij", job.structure, 3)
+    job.bonds = tuple(zip(src, dest))  
+    job.cns = np.bincount(src)
+    job.surface_indices = get_surface_indices(job.structure, job.cns)
+
+
+# # CO Placement
+# 
+# Place CO on surface, surface norm is in Z-direction
+
+# In[]:
+
+
+def place_co_molecule(job, site):
+    site_coords = job.structure[site].position
+    carbon_atom = ase.Atom("C", site_coords + [0, 0, 1.43])
+    oxygen_atom = ase.Atom("O", site_coords + [0, 0, 2*1.43])
+    carbon_monoxide = ase.Atoms((carbon_atom, oxygen_atom))
+    job.adsorption_states.append(job.structure + carbon_monoxide)
+    
+for job in jobs:
+    job.adsorption_states = []
+    for site in job.surface_indices:
+        place_co_molecule(job, site)
+
+
+# And finally, we'll freeze the surface to reduce the computational burden
+
+# In[]:
+
+
+def fix_all_slab_atoms(adsorption_state):
+    constraint = ase.constraints.FixAtoms(indices = [atom.index for atom in adsorption_state if atom.symbol not in ("C", "O")])
+    adsorption_state.set_constraint(constraint)
+
+for job in jobs:
+    for adsorption_state in job.adsorption_states:
+        fix_all_slab_atoms(adsorption_state)
+
+
+# # Adsorption Study
+# 
+# Eads=Ecomplex-(Eslab+Emolecule)
```

Comment:
This needs a more human-readable explanation

Reply:
Expanded the explanation